core/sr-staking-primitives/src/offence.rs

```diff
@@ -0,0 +1,147 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Common traits and types that are useful for describing offences for usage in environments
+//! that use staking.
+
+use rstd::vec::Vec;
+
+use codec::{Encode, Decode};
+use sr_primitives::Perbill;
+
+use crate::SessionIndex;
+
+/// The kind of an offence, is a byte string representing some kind identifier
+/// e.g. `b"im-online:offlin"`, `b"babe:equivocatio"`
+// TODO [slashing]: Is there something better we can have here that is more natural but still
+// flexible? as you see in examples, they get cut off with long names.
+pub type Kind = [u8; 16];
+
+/// Number of times the offence of this authority was already reported in the past.
+///
+/// Note we don't buffer offence reporting so everytime we see a new offence
+/// of the same kind, we will report past authorities again.
+/// This counter keeps track of how many times the authority was already reported in the past,
+/// so that we can slash it accordingly.
+pub type OffenceCount = u32;
+
+/// A type that represents a point in time on an abstract timescale.
+///
+/// See `Offence::time_slot` for details. The only requirement is that such timescale could be
+/// represented by a single `u128` value.
+pub type TimeSlot = u128;
+
+/// A trait implemented by an offence report.
+///
+/// This trait assumes that the offence is legitimate and was validated already.
+///
+/// Examples of offences include: a BABE equivocation or a GRANDPA unjustified vote.
+pub trait Offence<Offender> {
+	/// Identifier which is unique for this kind of an offence.
+	const ID: Kind;
+
+	/// The list of all offenders involved in this incident.
+	///
+	/// The list has no duplicates, so it is rather a set.
+	fn offenders(&self) -> Vec<Offender>;
+
+	/// The session index which is used for querying the validator set for the `slash_fraction`
+	/// function.
+	fn session_index(&self) -> SessionIndex;
+
+	/// Return a validators count at the time when the offence took place.
+	fn validators_count(&self) -> u32;
```

Comment:
`validator_set_count` might be a better name for this to imply that this count has to do with the validator set and not with the report.

Reply:
"Validators count" also isn't the right way to say that in english anyway.
"Validator count" would be fine

Reply:
Fixed.